### PR TITLE
ext_authz: do the authentication even the direct response is set

### LINKED ISSUE #17502: ext authz: the auth check is skipped when the direct response is set
*Title*: ext authz: the auth check is skipped when the direct response is set

*Description*:
https://github.com/envoyproxy/envoy/blob/a58fe011a399a545a1e247c114900b97db9cf1f2/source/extensions/filters/http/ext_authz/ext_authz.cc#L377

*Repro steps*:
```
    filter_chains:
    - filters:
      # The HTTP connection manager is the only network filter.
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  path: "/"
                direct_response:
                  status: 200
              typed_per_filter_config:
                envoy.filters.http.ext_authz:
                  "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthzPerRoute
                  disabled: true
          http_filters:
          - name: envoy.filters.http.ext_authz
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
              http_service:
                server_uri:
                  uri: 127.0.0.1:33334
                  cluster: auth_cluster
                  timeout: 0.25s
              with_request_body:
                max_request_bytes: 1024
                allow_partial_message: true
                pack_as_bytes: true
          - name: envoy.filters.http.router
 ```
Expect the envoy will call to auth_cluster to do the auth. But actually, it always returns 200 directly.

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: ext_authz: do the authentication even the direct response is set
Additional Description: 
This PR fixed the case the authentication will be skipped when the direct response is set.
And add the runtime flag `envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response` to enable the
revert of this behavior.

Also fix the merge config behavior, ensure the virtual host's typed config will be merged when
the direct response is set.

Risk Level: low
Testing: unittest 
Docs Changes: n/a
Release Notes: bug fixes note
Runtime guard: envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response
Fixes #17502
Related to #17377

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

### CODE REVIEW

**Comment by soulxu** [comment]
/wait

waiting for https://github.com/envoyproxy/envoy/pull/17449 merged

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 1]
```diff
@@ -374,7 +374,10 @@ void Filter::continueDecoding() {
 }
 
 Filter::PerRouteFlags Filter::getPerRouteFlags(const Router::RouteConstSharedPtr& route) const {
-  if (route == nullptr || route->routeEntry() == nullptr) {
+  if (route == nullptr ||
+      (!Runtime::runtimeFeatureEnabled(
```
Comment (dmitri-d) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:378]: Wouldn't this affect redirects too? Not sure if these be passed to ext_authz?

Comment (soulxu) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:378]: good point, I tested the redirect case, this change affect it.

@mattklein123 WDYT?

Comment (mattklein123) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:378]: Yeah I think it's correct to fix redirect also. You might want to rename the runtime flag and update the release notes?

/wait

Comment (soulxu) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:378]: right, let me update

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -42,6 +42,7 @@ Bug Fixes
 * access log: fix `%UPSTREAM_CLUSTER%` when used in http upstream access logs. Previously, it was always logging as an unset value.
 * aws request signer: fix the AWS Request Signer extension to correctly normalize the path and query string to be signed according to AWS' guidelines, so that the hash on the server side matches. See `AWS SigV4 documentaion <https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html>`_.
 * cluster: delete pools when they're idle to fix unbounded memory use when using PROXY protocol upstream with tcp_proxy. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.conn_pool_delete_when_idle`` runtime guard to false.
+* ext_authz: the authentication shouldn't be skpped when the direct response is set. This was fixed. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response_and_redirect`` runtime guard to false.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:45]:
```suggestion
* ext_authz: fixed skipping authentication when using returning either a direct response or a redirect. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response_and_redirect`` runtime guard to false.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with a small comment. Also needs a main merge.

If you are able to test a redirect case that would be great also.

/wait

File: docs/root/version_history/current.rst (hunk 2 above)
Comment (soulxu) [reply docs/root/version_history/current.rst]: thanks!

**Comment by soulxu** [comment]
> Thanks LGTM with a small comment. Also needs a main merge.
> 
> If you are able to test a redirect case that would be great also.
> 
> /wait

I added integration test for redirect and direct response case.

The unittest cover those two cases since both of them return empty routeEntry
https://github.com/envoyproxy/envoy/pull/17546/files#diff-64a51bc8f4c7ce4f31a3420225e6fd4d0f14a187424148da4b9dc87e11d97c96L1607

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -42,6 +42,7 @@ Bug Fixes
 * access log: fix ``%UPSTREAM_CLUSTER%`` when used in http upstream access logs. Previously, it was always logging as an unset value.
 * aws request signer: fix the AWS Request Signer extension to correctly normalize the path and query string to be signed according to AWS' guidelines, so that the hash on the server side matches. See `AWS SigV4 documentaion <https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html>`_.
 * cluster: delete pools when they're idle to fix unbounded memory use when using PROXY protocol upstream with tcp_proxy. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.conn_pool_delete_when_idle`` runtime guard to false.
+* ext_authz: fixed skipping authentication when using returning either a direct response or a redirect. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response_and_redirect`` runtime guard to false.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:45]:
```suggestion
* ext_authz: fixed skipping authentication when returning either a direct response or a redirect. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response_and_redirect`` runtime guard to false.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Awesome, thanks. Sorry just one more release note nit.

/wait

File: docs/root/version_history/current.rst (hunk 3 above)
Comment (soulxu) [reply docs/root/version_history/current.rst]: thanks, done

**Comment by mattklein123** [comment]
Sorry I think you need another main merge. :(

/wait

**Comment by soulxu** [comment]
> Sorry I think you need another main merge. :(
> 
> /wait

haha, no worries, let me do a main merge

**Review by mattklein123 — APPROVED** [review]

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -42,6 +42,7 @@ Bug Fixes
 * access log: fix ``%UPSTREAM_CLUSTER%`` when used in http upstream access logs. Previously, it was always logging as an unset value.
 * aws request signer: fix the AWS Request Signer extension to correctly normalize the path and query string to be signed according to AWS' guidelines, so that the hash on the server side matches. See `AWS SigV4 documentaion <https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html>`_.
 * cluster: delete pools when they're idle to fix unbounded memory use when using PROXY protocol upstream with tcp_proxy. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.conn_pool_delete_when_idle`` runtime guard to false.
+* ext_authz: fixed skipping authentication when returning either a direct response or a redirect. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.http_ext_authz_do_not_skip_direct_response_and_redirect`` runtime guard to false.
```
Comment (ggreenway) [inline docs/root/version_history/current.rst:45]: I think this should also be mentioned in "incompatible behavior changes". This behavior has been present for at least a few years, and I know some people are relying on the old behavior.

Comment (soulxu) [reply docs/root/version_history/current.rst:45]: that make sense, I can move it.

**Pull request merged**